File: MultiLanguageSandbox/src/test/deps/ruby/45.rb


# Finds M positive integers whose sum is N, such that their product is maximized.
# The function returns an array of these integers in lexicographically smallest order.
#
# Args:
#   n: A positive integer representing the sum of the integers to find.
#   m: A positive integer representing the number of integers to find.
#
# Returns:
#   An array of integers that are the lexicographically smallest set of integers
#   whose sum is N and whose product is maximized.
#
# Examples:
#   >>> max_product_partition(6, 3)
#   [2, 2, 2]
#   >>> max_product_partition(8, 3)
#   [2, 3, 3]


def max_product_partition(n, m)
  return [] if n < m || m <= 0

  quotient = n / m
  remainder = n % m

  # The optimal partition is (quotient + 1) repeated 'remainder' times,
  # followed by 'quotient' repeated (m - remainder) times.
  # But to get lex smallest, we need to have the smaller numbers first.
  # So the array should be [quotient] * (m - remainder) + [quotient + 1] * remainder
  # But since quotient + 1 > quotient, the lex smallest is when the smaller numbers come first.
  # So the correct order is [quotient] (m - remainder times), then [quotient + 1] (remainder times).

  result = []
  (m - remainder).times { result << quotient }
  remainder.times { result << quotient + 1 }

  result
end

raise 'Test failed' unless max_product_partition(6, 3) == [2, 2, 2]
raise 'Test failed' unless max_product_partition(8, 3) == [2, 3, 3]
raise 'Test failed' unless max_product_partition(10, 2) == [5, 5]
raise 'Test failed' unless max_product_partition(7, 3) == [2, 2, 3]
raise 'Test failed' unless max_product_partition(20, 5) == [4, 4, 4, 4, 4]

File: MultiLanguageSandbox/src/test/deps/php/45.php
<?php


/**
 * Finds M positive integers whose sum is N and whose product is as large as possible.
 * Outputs the combination with the smallest lexicographical order if there are multiple possibilities.
 * The function aims to distribute the total sum N into M parts such that the product of these parts is maximized.
 * 
 * Input: Two integers, $N representing the total sum and $M representing the number of parts.
 * Output: An array of integers representing the distribution that maximizes the product.
 * Example: Given N = 6 and M = 3, the function returns [2, 2, 2] as dividing 6 into three parts of 2 each maximizes the product 2*2*2 = 8.
 */

function maxProductDistribution($N, $M) {

    if ($M > $N) {
        return []; // Not possible to distribute N into M positive integers
    }
    
    $base = intdiv($N, $M);
    $remainder = $N % $M;
    
    $result = array();
    for ($i = 0; $i < $M; $i++) {
        if ($i < $remainder) {
            $result[] = $base + 1;
        } else {
            $result[] = $base;
        }
    }
    
    // To ensure lexicographical order, we need the smallest numbers first, but in this approach, the remainder parts are the larger numbers.
    // However, since we are placing the (base + 1) parts first, the array is already in non-decreasing order, which is the smallest lexicographical order for this scenario.
    // For example, for N=6, M=3: base=2, remainder=0 → [2,2,2].
    // For N=7, M=3: base=2, remainder=1 → [3,2,2], which is lex smallest.
    
    return $result;
}
function main(){
    // Test Case 1: Simple distribution
    assert(maxProductDistribution(6, 3) === [2, 2, 2]);

    // Test Case 2: When distribution is not perfectly even
    assert(maxProductDistribution(7, 3) === [2, 2, 3]);

    // Test Case 3: Larger numbers with an even distribution
    assert(maxProductDistribution(10, 2) === [5, 5]);

    // Test Case 4: Larger numbers with a remainder
    assert(maxProductDistribution(9, 4) === [2, 2, 2, 3]);
    assert(maxProductDistribution(9, 3) === [3,3, 3]);

    // Test Case 5: Single part (edge case)
    assert(maxProductDistribution(5, 1) === [5]);

}


main();
?>